src/core/job.c

```diff
@@ -1668,6 +1611,16 @@ int job_compare(Job *a, Job *b, UnitDependency assume_dep) {
         if (a->ignore_order || b->ignore_order)
                 return 0;
 
+        /* Figure out the dependency between units */
+        if (assume_dep == 0) {
+                if (hashmap_contains(ua->dependencies[UNIT_AFTER], ub))
+                        assume_dep = UNIT_AFTER;
+                /* We should not be called with a loop between jobs, let UNIT_BEFORE override not to fail. */
```

Comment:
In the case there would by a cycle between two units: A.After=B, B.After=B, then comparing the jobs for the order is not well defined (it's both >0 and <0, jobs aren't independent). Instead of failing hard, we simply ignore one of the deps and let the other win. However, in the current code, there should be no callers that would need to compare two jobs between whose units exists this cycle.

Reply:
OK. So maybe say:
> /* If we are called with a loop between jobs (a cycle  A.After=B, B.After=B exists), then order of the jobs is not well defined. We don't expect to be called in that case, but for robustness, instead of failing, ignore one of the deps and let the other win. */

Reply:
Done.